lib/Conversion/SCFToCalyx/SCFToCalyx.cpp

```diff
@@ -47,9 +47,44 @@ struct WhileScheduleable {
   calyx::GroupOp initGroup;
 };
 
-// A variant of types representing scheduleable operations.
+/// A variant of types representing scheduleable operations.
 using Scheduleable = std::variant<calyx::GroupOp, WhileScheduleable>;
 
+/// A structure representing a set of ports which act as a memory interface.
+struct CalyxMemoryPorts {
+  Value readData;
+  Value done;
+  Value writeData;
+  SmallVector<Value> addrPorts;
+  Value writeEn;
+};
+
+/// CalyxMemoryInterface defines an interface for accessing either a
+/// calyx::MemoryOp or a CalyxMemoryPorts struct.
+struct CalyxMemoryInterface {
+  CalyxMemoryInterface() {}
+  explicit CalyxMemoryInterface(const CalyxMemoryPorts &ports) : impl(ports) {}
+  explicit CalyxMemoryInterface(calyx::MemoryOp memOp) : impl(memOp) {}
+
+#define memoryInterfaceGetter(portName, TRet)                                  \
+  TRet portName() {                                                            \
+    if (auto memOp = std::get_if<calyx::MemoryOp>(&impl); memOp)               \
+      return memOp->portName();                                                \
+    else                                                                       \
+      return std::get<CalyxMemoryPorts>(impl).portName;                        \
+  }
+
+  memoryInterfaceGetter(readData, Value);
+  memoryInterfaceGetter(done, Value);
+  memoryInterfaceGetter(writeData, Value);
+  memoryInterfaceGetter(writeEn, Value);
+  memoryInterfaceGetter(addrPorts, ValueRange);
+#undef memoryInterfaceGetter
+
+private:
+  std::variant<calyx::MemoryOp, CalyxMemoryPorts> impl;
+};
```

Comment:
Can you explain your reasoning behind this, versus just using the `MemoryOp` itself, and its corresponding port helpers directly?

Reply:
Ah so if I understand correctly you're essentially "re-constructing" a memory from the ports passed in. IMO some reasoning for this would be helpful.

Reply:
> you're essentially "re-constructing" a memory from the ports passed in.

Correct. Will add comment, to clarify the intention.